amazon/ionbenchmark/ion_benchmark_cli.py

```diff
@@ -0,0 +1,360 @@
+
+    return file_size, result_with_gc
+
+
+def write_micro_benchmark_iterator(iterations, warmups, c_ext, obj, file, binary, profiling):
+    file_size = Path(file).stat().st_size / BYTES_TO_MB
+
+    # GC refers to reference cycles, not reference count
+    setup_with_gc = generate_simpleion_setup(gc=True, c_ext=c_ext)
+
+    test_code = generate_simpleion_dump_test_code(obj, profiling, binary=binary)
+
+    # warm up
+    timeit.timeit(stmt=test_code, setup=setup_with_gc, number=warmups)
+
+    # iteration
+    result_with_gc = timeit.timeit(stmt=test_code, setup=setup_with_gc, number=iterations) / iterations
+
+    return file_size, result_with_gc
+
+
+def write_micro_benchmark_event(iterations, warmups, c_ext, obj, binary, file=None):
+    pass
+
+
+def write_micro_benchmark_and_profiling(write_micro_benchmark_function, iterations, warmups, obj, c_ext, binary, file):
+    if not obj:
+        raise Exception("Invalid obj: object can not be none.")
+    if not write_micro_benchmark_function:
+        raise Exception("Invalid micro benchmark function: micro benchmark function can not be none.")
+    # Memory Profiling
+    write_micro_benchmark_function(iterations=1, warmups=0, obj=obj, c_ext=c_ext, file=file, binary=binary, profiling=True)
+
+    # Performance Benchmark
+    file_size, result_with_gc = \
+        write_micro_benchmark_function(iterations=iterations, warmups=warmups, obj=obj, c_ext=c_ext, file=file,
+                                       binary=binary, profiling=False)
+
+    # generate report
+    write_generate_report(file_size, result_with_gc, write_memory_usage_peak)
+
+    return file_size, result_with_gc, write_memory_usage_peak
+
+
+def write_generate_report(file_size, total_time, memory_usage_peak):
+    table = [['file_size (MB)', 'total_time (s)', 'memory_usage_peak (MB)'],
+             [format_decimal(file_size),
+              format_decimal(total_time),
+              format_decimal(memory_usage_peak)]]
+    print('\n')
+    print(tabulate(table, tablefmt='fancy_grid'))
+
+
+def ion_python_benchmark_cli(arguments):
+    if arguments['--version'] or arguments['-v']:
+        print(TOOL_VERSION)
+        return TOOL_VERSION
+    elif arguments['generate']:
+        print('Generate feature is not supported yet')
+        return 0
+    if not arguments['<input_file>']:
+        raise Exception('Invalid input file')
+    file = arguments['<input_file>']
+    iterations = int(arguments['--iterations'])
+    warmups = int(arguments['--warmups'])
+    c_ext = str_to_bool(arguments['--c-extension'])
+    binary = str_to_bool(arguments['--binary'])
+
+    if arguments['read']:
+        api = arguments['--api']
+        if not api or api == API.SIMPLE_ION.value:
+            read_micro_benchmark_function = read_micro_benchmark_simpleion
+        elif api == API.ITERATOR.value:
+            read_micro_benchmark_function = read_micro_benchmark_iterator
```

Comment:
`iterator` option has the same underlying algorithm as `simpleion`. I deprecated.

Reply:
There is a difference in performance characteristics between iterator mode enabled and disabled. For iterator mode, you don't need to hold all top-level values in memory, while when disabled you do. This can make a big difference for large streams. We should definitely have an option to try both modes.